ecscollector/collector_test.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2025 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecscollector
+
+import (
+	"bytes"
+	_ "embed"
+	"flag"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/prometheus-community/ecs_exporter/ecsmetadata"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Create a metadata client that will always receive the given fixture API
+// responses.
+func fixtureClient(taskMetadata, taskStats []byte) (*ecsmetadata.Client, *httptest.Server) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("GET /task", func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("content-type", "application/json")
+		w.Write(taskMetadata)
+	})
+	mux.HandleFunc("GET /task/stats", func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("content-type", "application/json")
+		w.Write(taskStats)
+	})
+
+	server := httptest.NewServer(mux)
+	return ecsmetadata.NewClient(server.URL), server
+}
+
+// Renders ecs_exporter metrics from the given metadata client to the prometheus
+// text exposition format.
+func renderMetrics(client *ecsmetadata.Client) ([]byte, error) {
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(NewCollector(client, slog.Default()))
+
+	// It seems that the only way to really get full /metrics output is with
+	// promhttp.
```

Comment:
Would this package help? https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/testutil#CollectAndCompare

Reply:
Thanks, I did not know about this package.  It looks like I could make this work with CollectAndCompare and CollectAndFormat (to update the snapshots).  I will try it out.  

Do you think it's preferable to use these functions in general, instead of the approach I took?

Reply:
Ah, strangely enough they have opposite behaviors if you pass no `metricNames`, which is what I want to do: CollectAndCompare drops no metrics, while CollectAndFormat drops all metrics (thus emitting the empty slice).  Well.

Reply:
I looked into using [Describe](https://github.com/prometheus-community/ecs_exporter/blob/a72e49c296b73def69187bbecc655f884f9ef656/ecscollector/collector.go#L169) to automate getting the metricNames to use in these testutil functions, but the names are private on Desc.  So I'd either have to expose the metric names explicitly from the collector to make these tests work, or I'd have to hard code them in the tests.

I'm not sure there's a point to doing that over what I've already got.  What do you think?

Reply:
This project doesn't have snapshots today, so I wouldn't be too bothered about having a way to automatically update the snapshots (like with the flag added below here). With that said, I think CollectAndCompare would serve to replace the snapshot comparison. I do that in most of my exporter tests i.e. https://github.com/prometheus-community/elasticsearch_exporter/blob/master/collector/slm_test.go#L30. 

I'm not one of the usual maintainers on this project though so they might have a different opinion.

Reply:
Okay. I can try CollectAndCompare only, and my httptest-based way to do automated snapshot updates.  CollectAndCompare certainly gives a better error message than my current setup.

> I'm not one of the usual maintainers on this project
There is no usual maintainer on this project, just @SuperQ's charity 🌞.  I would like to be that maintainer.

Reply:
Done.